### DISCUSSION TITLE
transient queue can disappear that used to be on a node that went down and gets successfully declared, bind and start consuming from on a new node

### DISCUSSION THREAD

**Discussion by Jaqy** [rabbitmq/rabbitmq-server#6272] [Other]
We have an issue where a transient queue can disappear that used to be on a node that went down and gets successfully declared, bind and start consuming from on a new node without the consumer knowing that the queue is gone.

From what I have gathered, this seems to happen between when the running nodes detect that a node is down with logs: **“rabbit on node ‘[vhost-address-of-node-that-went-down]' down”** 

and when the running nodes have finished deleting old transient queues from the node that went down with log: **“77 transient queues from an old incarnation of node '[vhost-address-of-node-that-went-down]' deleted in 2.627854s”**

And any transient queues declared on the running nodes, that used to be on the node that went down, have a chance of disappearing. 

Looking at the rabbitmq code and tracing down the source of the **“77 transient queues from an old incarnation of node '[vhost-address-of-node-that-went-down]' deleted in 2.627854s”** log: 

rabbit_amqqueue.erl:1998
```
on_node_down(Node) ->
    {Time, {QueueNames, QueueDeletions}} = timer:tc(fun() -> delete_queues_on_node_down(Node) end),
    case length(QueueNames) of
        0 -> ok;
        _ -> rabbit_log:info("~p transient queues from an old incarnation of node ~p deleted in ~fs", [length(QueueNames), Node, Time/1000000])
    end,
    notify_queue_binding_deletions(QueueDeletions),
    rabbit_core_metrics:queues_deleted(QueueNames),
    notify_queues_deleted(QueueNames),
    ok.
```

Going to: 
rabbit_amqqueue.erl:2009
```
delete_queues_on_node_down(Node) ->
    lists:unzip(lists:flatten([
        rabbit_misc:execute_mnesia_transaction(
          fun () -> [{Queue, delete_queue(Queue)} || Queue <- Queues] end
        ) || Queues <- partition_queues(queues_to_delete_when_node_down(Node))
    ])).
```

It seems that we first get all the queues to be deleted in our mnesia table with: `queues_to_delete_when_node_down()` (rabbit_amqqueue.erl:2033)

And afterwards we start deleting those queues in batches of 10 queues at the time with this `partition_queues()` (rabbit_amqqueue.erl:2028) function and the `delete_queue() `function just deleting the queues without checking on which node the queue belongs to. 

Thus it seems possible to create a queue with the same name that used to be on the node that went down between the deletion of queues partitions and then getting deleted again in the following batch of queue deletions. 

And because this on_node_down logic is assuming that all these queues are not alive anymore, the rabbitmq doesn’t send a `basic_cancel` to the consumers and the consumer never knows that the queue is deleted. 

With my very limited knowledge of the rabbitmq codebase, I am wondering if it would be possible to ensure that the queue we delete in the mnesia table still belongs to the node that went down in the same mnesia transaction? Or maybe there is a another solutions to solve this?

**Comment by michaelklishin** [comment]
There is a natural race condition between what clients and nodes do in response to a node failure.
Use a quorum queue or a stream if you need better data safety guarantees. A transient
non-replicated queue won't offer much.

**Reply by Jaqy** [reply]
Quorum queues are to slow for our requirements.

And from what I could tell from the on_node_down() process the node does not send any notification to the client of the queues that get deleted. 

It is not data safety we are looking for, but die fast and recover fast.

**Reply by Jaqy** [reply]
But is my assessment of that you can create a queue in between the batch deletetion of queues and have a risk of your created queue being deleted again in a following batch wrong? 

Because if my assessment is correct, I don't understand why this isn't considered a issue then?

**Reply by michaelklishin** [reply]
In modern versions quorum queues are more efficient than classic queues, not less. Streams are even more efficient. Transient queue efficiency isn't meaningfully better beyond a certain backlog. In 3.12 and beyond,
the idea of a transient entity becomes a moot point, and in 4.0, transient entities will cease to exist entirely.

@lhoguin suggested that we can skip deletion of queues that have a live replica PID on a connected node. That would help with one concurrency hazard (but not every possible one).

**Reply by mkuratczyk** [reply]
@Jaqy could you share your workload details? Ideally as a [perf-test command](https://perftest.rabbitmq.com/), but if you are not familiar with it, as much detail as you can about the number of queues, publishers, consumers message throughput, message size, etc. As Michael said, there is a good chance the alternatives are already more performant but I'm always interested in learning about real-world cases, so that we can make sure we can handle them well (especially as we want to deprecate some features). Also what is your definitions of performance - do you have a latency target of some kind for example? Thanks,

**Reply by Jaqy** [reply]
We are continously performance testing quorum queues and we see an overall latency increase of around 30% whe  we replace our queues with quorum queues. And I believe we just upgraded to the latest version. And this increase is regardless of load we put on it. 

We haven't tested stream queues yet. 

But are you saying that in 4.0, we can only declare a stream queue or quorum queue? 

I don't have access to all those details right now, but I will provide them tomorrow.

**Reply by mkuratczyk** [reply]
No, classic queues v2 will also be available (but without mirroring in 4.0). We have "parts" of them in 3.10/3.11 and "final" version on the main branch (of course it's still being tested and can be modified here and there but we believe they are pretty much ready).

I'd really appreciate the details as we obviously run all kinds of tests and if you say you are consistently getting worse results with newer queue types/version, that's something we'd like to look into. Thank you

**Reply by michaelklishin** [reply]
Any Raft-based replicated queue type will have a somewhat higher latency. However, throughput and throughput variability of quorum queues and streams can (and in practice often is) be higher than even that of non-mirrored classic queues.

CQv2's main benefit is memory footprint stability, not latency reduction.

**Reply by Jaqy** [reply]
@mkuratczyk 

We are running currently with version: `3.11.2`, I couldn't do the perf-test command, but currently this is what we have running in our staging environment, which is running the heaviest load atm. 

We have a bunch of microservices that use RabbitMQ to communicate with each other. 
We got 3 node RabbitMQ cluster
~496 queues 
- ~111 reply queues
- ~300 CQv2 queues (not replicated) that expires and are exclusively consumed
- ~85 CQv2 queues (not replicated) with a 1000-1500 message-ttl and on avg have 6-9 consumers

~103 connections
~1291 channels
~98 exchanges
~1269 Consumers
~ Publish rate 3000s - 2500/s
~ Deliver (manual ack) & consumer ack ~2000/s - 2800/s
~ Deliver (auto ack) ~500-700/s
I am not sure about the message size, but it's small. The memory footprint in each of the 3 nodes with a 2500/s publish rate is around 450 - 500 MiB 
Generally messages are consumed as soon as they arrive in the queue.


When I tested quorum queues, I replaced these ~85 CQv2 with 3 replica QQ’s and our latency went from `~32ms` to `~45ms` and that time is from when our services received a request from the outside to send a response back. This I tested with RabbitMQ version 3.10.0. But currently a colleague is testing QQ's again with 3.11 and we still see this latency increase.

**Reply by kjnilsson** [reply]
> When I tested quorum queues, I replaced these ~85 CQv2 with 3 replica QQ’s and our latency went from ~32ms to ~45ms and that time is from when our services received a request from the outside to send a response back. This I tested with RabbitMQ version 3.10.0. But currently a colleague is testing QQ's again with 3.11 and we still see this latency increase.

There won't be anything magically reducing QQ latency in 3.11. Quorum queue messages are always fsynced to disk before being publisher confirms are issued or consumer deliveries are done which is probably where most of the latency is picked up. What kind of disks do you use?

**Reply by Jaqy** [reply]
@kjnilsson I don't have access to the infrastructure, so I can't tell what specific disk is used, but it is premium storage. 

But regardless a non replicated CQ will always be faster then a replicated QQ because of the writing to disk. We prioritize speed over data safety, and therefor are these replicated queues not an option for us. 

The latency increase of one message passing through via a QQ is probably small, but our latency performance increase is a accumulation of messages passing through the rabbit multiple times.

**Reply by haljin** [reply]
> @kjnilsson I don't have access to the infrastructure, so I can't tell what specific disk is used, but it is premium storage.

The disk is `managed-premium` which in Azure maps to [Premium LRS (SSD)](https://azure.microsoft.com/en-us/pricing/details/managed-disks/). It's entirely possible that Azure is simply throttling our IO operations, since we request a rather small disk - we really do not need the storage and persistence at all. Our use case is more about latency.

We've also seen funny patterns where at some times of the day the quorum queues become slower by an order of magnitude and then recover again, probably as some other workloads kick in in our K8s cluster. 

(I am on @Jaqy 's team)

**Reply by kjnilsson** [reply]
I wouldn't normally suggest this but have you considered the rabbitmq-sharding plugin at all? It has some downsides which may be blockers but in your case it seems like you want the queues to be available even when nodes are down and the sharded queue will always have a "shard" on each node.

**Reply by kjnilsson** [reply]
If you wanted to test the latency of quorum queues _without_ the mandatory fsync operation that is possible by configuring the `wal_sync_method=none` in the `ra` application. You'd have to use the advanced config option for this and isn't something we'd recommend but it may be an interesting data point.

**Reply by mkuratczyk** [reply]
Here's what we see in our tests. This is not exactly what you are doing (I will try to build a test that closer mimics your workload in the coming weeks) but CQv2 latency is really good. The workload here was 100 queues, each doing 100 msgs/s with 1 publisher and 1 consumer. So a total of 10k msgs/s on a single-node RabbitMQ (deployed to GKE, 14 fully utilized CPU cores, 12GB RAM, hardly used at all).

```
perf-test -x 100 -y 100 -c 10 -s SIZE -qp q-%d -qpf 1 -qpt 100 -qa x-max-length=1000000 -ad false -f persistent -qa x-queue-version=2 -r 1000 # queue args adjusted for the queue type
```

I compared CQv1, CQv1 lazy, CQv2 3.11, CQv2 from the `main` branch.

Here are the results with 12 byte messages:
![size-12](https://user-images.githubusercontent.com/9566114/198633077-741e1e1a-55bd-4d78-b6e8-f2f107ccdba8.png)

100 byte messages:
![size-100](https://user-images.githubusercontent.com/9566114/198633268-82f841af-8d09-4726-87bb-d2f76f16098f.png)

1000 byte messages:
![size-1000](https://user-images.githubusercontent.com/9566114/198633429-93262685-306e-4b5d-a5a5-4127b2512b78.png)

(I didn't have time to investigate those spikes but CQv2 doesn't seem affected, even in 3.11).

**Reply by Jaqy** [reply]
@mkuratczyk 

CQv2 queues are very for performance, and we like to continue using them. 

The problem we are facing with CQv2's is that there is this race condition where there is a chance they can get deleted when a node goes down and the remaining nodes go through this `on_node_down()` process and they used to be on the node that went down, and the consumer declares it on one of the remaining nodes, due to this batch queue deletion: 
```
on_node_down(Node) ->
    {Time, {QueueNames, QueueDeletions}} = timer:tc(fun() -> delete_queues_on_node_down(Node) end),
    case length(QueueNames) of
        0 -> ok;
        _ -> rabbit_log:info("~tp transient queues from an old incarnation of node ~tp deleted in ~fs", [length(QueueNames), Node, Time/1000000])
    end,
    notify_queue_binding_deletions(QueueDeletions),
    rabbit_core_metrics:queues_deleted(QueueNames),
    notify_queues_deleted(QueueNames),
    ok.

delete_queues_on_node_down(Node) ->
    lists:unzip(lists:flatten([
        rabbit_misc:execute_mnesia_transaction(
          fun () -> [{Queue, delete_queue(Queue)} || Queue <- Queues] end
        ) || Queues <- partition_queues(queues_to_delete_when_node_down(Node))
    ])).

delete_queue(QueueName) ->
    ok = mnesia:delete({rabbit_queue, QueueName}),
    rabbit_binding:remove_transient_for_destination(QueueName).

% If there are many queues and we delete them all in a single Mnesia transaction,
% this can block all other Mnesia operations for a really long time.
% In situations where a node wants to (re-)join a cluster,
% Mnesia won't be able to sync on the new node until this operation finishes.
% As a result, we want to have multiple Mnesia transactions so that other
% operations can make progress in between these queue delete transactions.
%
% 10 queues per Mnesia transaction is an arbitrary number, but it seems to work OK with 50k queues per node.
partition_queues([Q0,Q1,Q2,Q3,Q4,Q5,Q6,Q7,Q8,Q9 | T]) ->
    [[Q0,Q1,Q2,Q3,Q4,Q5,Q6,Q7,Q8,Q9] | partition_queues(T)];
partition_queues(T) ->
    [T].

queues_to_delete_when_node_down(NodeDown) ->
    rabbit_misc:execute_mnesia_transaction(fun () ->
        qlc:e(qlc:q([amqqueue:get_name(Q) ||
            Q <- mnesia:table(rabbit_queue),
                amqqueue:qnode(Q) == NodeDown andalso
                not rabbit_mnesia:is_process_alive(amqqueue:get_pid(Q)) andalso
                (not amqqueue:is_classic(Q) orelse not amqqueue:is_durable(Q)) andalso
                (not rabbit_amqqueue:is_replicated(Q) orelse
                rabbit_amqqueue:is_dead_exclusive(Q))]
        ))
    end).
```

But said that, I am happy to see that you guys have taken this issue up #6274, and I am looking forward to this getting solved :)

**Reply by Jaqy** [reply]
> If you wanted to test the latency of quorum queues _without_ the mandatory fsync operation that is possible by configuring the `wal_sync_method=none` in the `ra` application. You'd have to use the advanced config option for this and isn't something we'd recommend but it may be an interesting data point.

That would be interesting and I have forwarded it to my team member how is currently doing some QQ performance testing

**Reply by Jaqy** [reply]
> I wouldn't normally suggest this but have you considered the rabbitmq-sharding plugin at all? It has some downsides which may be blockers but in your case it seems like you want the queues to be available even when nodes are down and the sharded queue will always have a "shard" on each node.

So far I know, we haven't. But maybe it is something we could consider? @haljin

**Reply by mkuratczyk** [reply]
I understand the issue, that part is clear. I just wanted to take this opportunity to learn about your workload, since we rarely have a chance to talk to a user who is "continuously performance testing quorum queues", and the fact that you see a performance drop was concerning. But I think we understand what you do now and you gave me some ideas to add to our benchmarking tests as some areas are not covered to the extent they can. Thanks

**Reply by Jaqy** [reply]
@mkuratczyk ah ok, I just wanted to make sure that you weren't wasting any time on my behalf. You can never run enough performance tests :)

**Reply by mkuratczyk** [reply]
@Jaqy any chance you can share your testing method/tool? We do most of the tests with perf-test, which currently doesn't support Direct Reply-To. We may add this in the future but if you have a tool I could use to test direct reply-to performance, that would be great.

Also, feel free to reach out to me with anything related to performance testing (testing process, results potential bottlenecks). You will easily find me on the RabbitMQ slack. I'd love to learn more about your use cases and issues and hopefully I can help a bit too. :)

Thanks,

**Comment by michaelklishin** [comment]
I have filed a specific issue with a couple of ideas in mind https://github.com/rabbitmq/rabbitmq-server/issues/6274.

A more attractive option for most people would be switch to quorum queues or streams with the smallest number of supported replicas (three). Channel operations in flight when a new leader
election happens will be delayed until after a leader is available. So clients don't have to do anything. On the other hand, latency inevitably will increase for that brief moment.

**Reply by Jaqy** [reply]
@michaelklishin 

Unfortunately are QQ's not an option for us in our use case, so I am very happy to see you have raised the issue again.
